lib/ui/painting/codec.cc

```diff
@@ -17,6 +17,8 @@
 #include "third_party/tonic/logging/dart_invoke.h"
 #include "third_party/tonic/typed_data/uint8_list.h"
 
+#define fuzzyEquals(x, y) (fabs(x - y) < DBL_EPSILON)
```

Comment:
This concerns me a bit, do we really need to resize if the size is scaled by 5%? Does skia allow doing this automatically?

Reply:
Skia has mipmap which allows fast resizing to any size at a memory cost of ~1.3x. I think @chinmaygarde has done some experiments with it in the engine. Chinmay: what's your results and current status of using mipmap?

Reply:
BTW, I think `fuzzyEquals` probably doesn't follow our naming style. All the `#define` I see has a name like `FOO_BAR`. See go/fluttercpp (which redirects to https://google.github.io/styleguide/cppguide.html)

Reply:
I was hoping that someone would point me to an already existing implementation and I could delete this from here. Something similar to https://chromium.googlesource.com/angle/angle/+/master/src/common/mathutil.h, where I could just use `AlmostEquals`. I will rename this to `FUZZY_EQUALS`.

Reply:
Oh, I just saw https://github.com/flutter/flutter/issues/26194 which is about making thumbnails to avoid OOM. Mipmap won't help OOM.

In this case, I wouldn't worry about `fuzzyEquals` at all. If the developer asks to resize, let's always resize even if they differ just by 1 pixel. Otherwise, it will be very confusing: I'll probably file a bug if I asks Flutter to resize to (500, 500) but instead get an image of (501, 501). (Note that Skia's image only has integer width/height, so it doesn't make sense to me to take a double width/height input.)

If they're worried about the performance (decoding and resizing time), they can turn off resizing by checking the image size in their own code instead of us doing something opaque in the engine. Do we currently have an API to check image size without fully decoding? If not, we can probably add that to solve the problem.

BTW, here's another worry if developers can't know the original width/height: how could developers preserve the width/height ratio using this API if they don't know the original width/height?

Reply:
We don't have an API to read image size without decoding the full image.